ext/winevt/winevt_channel.c

```diff
@@ -253,8 +276,15 @@ rb_winevt_channel_each(VALUE self)
       rb_raise(rb_eRuntimeError, errBuf);
     }
 
-    status = is_subscribable_channel_p(hChannelConfig, winevtChannel->force_enumerate);
-    if (status != ERROR_SUCCESS)
+    struct subscribable_channel_data data = {
+      hChannelConfig,
+      winevtChannel->force_enumerate
+    };
+
+    is_subscribable = rb_ensure(
```

Comment:
`buffer` also leaks.

Reply:
Fixed in https://github.com/fluent-plugins-nursery/winevt_c/pull/23/commits/57a301f7adc420351411b66163953df17ca49f02.

Reply:
`winevtChannel->channels` should be also closed since surrounding codes do it.

Reply:
oh!

Reply:
Really? Executing `EvtClose(winevtChannel->channels)` in the latter `if(!is_subscribable)` causes ErrorCode 6.

Reply:
Oops, you are right.
Freeing buffer isn't also needed.

Reply:
I see.... 😅

Reply:
> Oops, you are right.
> Freeing buffer isn't also needed.

No, it's not correct.
They (`winevtChannel->channels` -> `buffer`) aren't freed in error cases since error is still raised by `rb_ensure()`.

I think `is_subscribable_channel_p()` shouldn't raise errors to avoid confusion, they should be raised at caller.

Reply:
Hmm...., error handling is so difficult.... :sweat:

Reply:
How about this? https://github.com/fluent-plugins-nursery/winevt_c/pull/23/commits/cdf779ba9520076913cd59f0e0c26a6b6a9a6498